### ISSUE TITLE
Internet isolation for Docker container

### ISSUE THREAD

**Issue by midavow569** [Py-KMS-Organization/py-kms#55]
I suggest to add command parameter to isolate py-kms container from Internet to prevent even accidental data leaks or vulnerabilities exploitation.

I am not an expert in Docker, but I found the answer https://stackoverflow.com/a/42336592

Also the note that DNS is still available https://docs.docker.com/config/containers/container-networking/ and it is needed to find a solution to forbid the DNS for py-kms.

**UPD.** It is for local use.

**Comment by simonmicro** [comment]
**TL;DR** _No. This is not the scope of py-kms. Fix your firewall._

So, py-kms is listening to `0.0.0.0` by default (can be overwritten inside the container using the environment variable `IP`) - you can change the IPv4 e.g. to `127.0.0.1` using Docker (e.g. for the command line use ` -p 127.0.0.1:1688:1688` (I think) instead of ` -p 1688:1688`) if you do not trust py-kms.

Keep in mind Docker will answer IPv6 requests by using an internal IPv4 NAT - therefore py-kms won't be able to check if that attempt comes from "the internet". To really block requests from outside take a look into `iptables` and how to use them with Docker - this is not in the scope of py-kms.

> I am not an expert in Docker, but I found the answer https://stackoverflow.com/a/42336592

This idea is flawed, as this creates an internal network only for use inside Docker, the `--internal` parameter should prevent even the host from accessing it and furthermore any external Windows client. This networking mode is commonly only used in Swarm deployments anyways. In any case the network is not routed by you system and therefore not accessible even if the `--internal` parameter is being omitted.

> Also the note that DNS is still available https://docs.docker.com/config/containers/container-networking/ and it is needed to find a solution to forbid the DNS for py-kms.

I do not follow on this one... If you add a DNS entry for your py-kms server into your DNS server, it is accessible by all DNS clients. If you do not want this, you have to fix that inside the DNS server to not serve this entry to any client.

> UPD. It is for local use.

...what does UDP stand for?

> I suggest to add command parameter to isolate py-kms container from Internet to prevent even accidental data leaks or vulnerabilities exploitation.

In case you think py-kms would try to communicate with anything external, you can be assured it is not. Take a look into the sources, there are not calls to the outside included, as everything needed is embedded into the container itself. In case you want to ensure this, set a fixed IPv4 for the py-kms container and add a rule to `iptables` to prevent any outgoing new connection (make sure to allow outgoing established connections!).
